torch/_inductor/runtime/benchmarking.py

```diff
@@ -0,0 +1,944 @@
+        # Warm-up
+        for _ in range(n_warmup):
+            fn()
+        # Benchmark
+        for i in range(n_repeat):
+            # we don't want `fn` to accumulate gradient values
+            # if it contains a backward pass. So we clear the
+            # provided gradients
+            if grad_to_none is not None:
+                for x in grad_to_none:
+                    x.grad = None
+            # we clear the L2 cache before each run
+            cache.zero_()
+            # record time of `fn`
+            start_event[i].record()
+            fn()
+            end_event[i].record()
+        # Record clocks
+        torch.cuda.synchronize()
+        times = torch.tensor(
+            [s.elapsed_time(e) for s, e in zip(start_event, end_event)],
+            dtype=torch.float,
+        )
+        if quantiles is not None:
+            ret = torch.quantile(
+                times, torch.tensor(quantiles, dtype=torch.float)
+            ).tolist()
+            return ret[0]
+        return getattr(torch, return_mode)(times).item()
+
+    @time_and_log
+    @maybe_fallback_to_original_benchmarking("original_do_bench_many_gpu")
+    def benchmark_many_gpu(
+        self: Self,
+        callables: List[Callable[[], Any]],
+        estimation_iters: int = 5,
+        memory_warmup_iters: int = 100,
+        benchmark_iters: int = 100,
+        max_benchmark_duration_ms: int = 25,
+        ranking_key: Optional[str] = None,
+        pruning_key: Optional[str] = None,
+        pruning_factor: float = 1.1,
+        pruning_limit: int = 5,
+        **kwargs: Any,
+    ) -> List[float]:
+        # we don't want any outside errors propagating into benchmarking
+        torch.cuda.synchronize()
+
+        # initialize here, avoids double buffer allocation
+        self.cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear
+
+        # this will fail the entire group if any of the callables fail,
+        # in the case that users directly call this method they should
+        # handle this case on their end. in the case that we get here
+        # from a grouping of lazy benchmarks, we should never have failures
+        # since they will have already been checked previously
+        for _callable in callables:
+            _callable()
+        torch.cuda.synchronize()
+
+        buffer = torch.empty(self.L2_cache_size // 4, dtype=torch.int, device="cuda")
+        buffer.zero_()
+
+        interleaved_event_pairs = self.get_interleaved_event_pairs(
+            len(callables), estimation_iters
+        )
+        torch.cuda._sleep(int(1 / self.gpu_time_ms_per_gpu_clock_cycle))
+        start_time_s = perf_counter()
+        for event_pairs in interleaved_event_pairs:
+            for _callable, (start_event, end_event) in zip(callables, event_pairs):
+                buffer.zero_()
+                start_event.record()
+                _callable()
+                end_event.record()
+        cpu_launch_overhead_ms_per_iter = (
```

Comment:
I dont that its a valid assumption that all of the kernels will have a proportional cpu launch overhead

Reply:
But they should be *close enough*, the sleep duration doesn't have to be entirely perfect we just want to avoid sleeping for 20x choices when 15x were pruned